### DISCUSSION TITLE
LLM api backend with multiple replicas: Support routing request to same instance of backend

### DISCUSSION THREAD

**Discussion by asrinivasanTT** [open-webui/open-webui#6857] [Ideas]
In larger scale deployments of LLM APIs, multiple instances of vLLM/ollama can be deployed across multiple nodes in a cluster, and a HTTP load balancer is configured to route requests to the multiple instances of the backend. Backends like vLLM support KV-Cache reuse across successive requests from the same user under a concept called prefix caching. If the load balancer is configured to route all requests from the same user to the same backend instance, the response times for the user can be significantly lowered.

Commonly used HTTP load balancers like NGINX and Traefik support this kind of a routing and call it sticky sessions. They implement sticky sessions by adding in a session cookie to the response it sends back to the user. The user sends back this session cookie in subsequent requests and thus the load balancer is able to route the request to the backend instance that originally served the user.
1. NGINX - https://docs.nginx.com/nginx/admin-guide/load-balancer/http-load-balancer/#enabling-session-persistence
2. Traefik - https://doc.traefik.io/traefik/routing/services/#sticky-sessions

open-webui currently does not support this. It does not send any form of cookies/identification to the LLM backend when it makes a request. One potentially complex way to implement this would be for open-webui to track more user state and store the cookie it receives from the LLM backend per chat session. This would be very complex to implement.

What open-webui does already is that it sends any headers (and thus cookies) it receives from the backend to the user's browser. See [here](https://github.com/open-webui/open-webui/blob/96c865404d36637eafadb6d2dd2365c85d452648/backend/open_webui/apps/openai/main.py#L463). Thus, I propose a much simpler solution where adding an env var at runtime named `PASSTHROUGH_SESSION_COOKIES` that stores a comma separated list of cookie names to passthrough to the backend. If this var is set, any cookies matching this names in this list that it receives from the browser in subsequent requests will be forwarded on to the LM backend.

Someone has already requested a feature like this before [here](https://github.com/open-webui/open-webui/discussions/5109) and [here](https://github.com/open-webui/open-webui/issues/607)

I will be opening a PR for this soon. Comments are welcome

**Comment by av** [comment]
Having an ability to set and acess cookies while having WebUI backend as a proxy would be a super valuable feature for building stateful workflows for assistant and agent use-cases. 

I landed here investigating how to implement multi-turn conversations in a custom OpenAI-compatible server.

**Comment by tjbck** [comment]
Related: https://github.com/open-webui/open-webui/discussions/6999

You can most definitely pass user information back to the backend using a custom Function.

**Comment by av** [comment]
Yes, a function will have access to user metadata and can indeed supply any necessary extras in the payload or headers.

**Comment by asrinivasanTT** [comment]
I opened a PR last week. Please checkout https://github.com/open-webui/open-webui/pull/6858
